src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -873,14 +873,18 @@ public Map<Arena, List<Level>> getLevels(Collection<? extends CompactionSSTable>
         for (Arena arena : arenas)
         {
             List<Level> levels = new ArrayList<>(MAX_LEVELS);
-            arena.sstables.sort(currentShardManager::compareByDensity);
+            // Use a cache to avoid recomputing the density of the same sstable multiple times
+            Map<CompactionSSTable, Double> sstableDensityCache = new HashMap<>();
+
+            // Sort the list in place based on the expensive key
+            arena.sstables.sort(Comparator.comparing(sstable -> sstableDensityCache.computeIfAbsent(sstable, currentShardManager::density)));
```

Comment:
what about calling currentShardManager::density once outside of this sort ? otherwise the number of calls depends a little bit on the sorting algorithm

`arena.sstables.forEach(sstable -> sstableDensityCache.put(sstable, currentShardManager.density(sstable))`

if we precompute then all the calls to `computeIfAbsent` because simple `get`